flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/UrlDecodeFunction.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.functions.scalar;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.SpecializedFunction;
+
+import javax.annotation.Nullable;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#URL_DECODE}. */
+@Internal
+public class UrlDecodeFunction extends BuiltInScalarFunction {
+
+    public UrlDecodeFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.URL_DECODE, context);
+    }
+
+    public @Nullable StringData eval(StringData value) {
+        final Charset charset = StandardCharsets.UTF_8;
+        try {
+            return StringData.fromString(URLDecoder.decode(value.toString(), charset.name()));
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(
+                    "Failed to decode value: " + value + " with charset: " + charset.name(), e);
+        } catch (RuntimeException e) {
+            return value;
+        }
```

Comment:
why do we need that?

Reply:
we should return the original value if failed, just like Calcite/Spark does.

Reply:
@superdiaodiao
Do we know why Calcite/Spark do this? It seems to be a strange design choice as we will go forward with this string not decoded, but the subsequent code would expect it to be decoded and is likely to get unexpected results; which would be tricky to debug.

Reply:
In the Calcite, the author didn't demonstrate the reason and the origin text is: https://github.com/apache/calcite/pull/3318#discussion_r1372603727

While in the Spark, it is different with the Calcite, although the Calcite said that it can be served for Spark. In fact, Spark will throw an exception instead of returning the origin value in this case.

So what's Flink team's expectation? I will follow your guide.

Reply:
I would follow same approach as Spark does otherwise it's impossible to say whether the function failed or succeeded... 
Also we need a test for this case which seems to be absent ...

Reply:
ok, I will make it throwable and add test

Reply:
Compared with returning the original string, returning null can better confirm whether the parsing failed. The string functions in flink basically follow this specification. However, I have observed the implementation of url_decode implemented by users, and there are two implementations of returning the original string and null, so I have no particular preference for the choice here. cc @lincoln-lil

Reply:
I would vote for `null` since it will allow to understand, that the function was not able to encode it correctly

Reply:
> I would vote for `null` since it will allow to understand, that the function was not able to encode it correctly

You mean, these two funcs should return null when decode or encode failed?

Reply:
yep, probably it will be better if both return `null`

the problem with failure could be that if we have only one record failing the function it could fail the whole query...
I'm pretty sure this is not the behavior we want... With `null` we seems on a bit safer side
please correct me if i'm wrong

Reply:
There aren't many examples of this function in other technologies, but I think most of them don't fail but return `null`. I would also be fine with returning `null` during exceptions. Note: I think in practice it will be quite tricky to get into a failure state with this function.

Reply:
Yes, returning null makes funcs more stable, 100% agree.

Reply:
For me, I would be keener to have an error, it is then obvious this is an out of the ordinary situation and as @MartijnVisser says this is going to be extremely unlikely to occur. In this case we would be treating runtime exceptions like errors (e.g. out of memory). 
 
If we return a null then potentially we have some sort of null pointer or lack of integrity in subsequent processing due to the missing data. It would be tricky to understand that this is the root cause.
 
If the consensus is to return a null,
- would we document that null is returned for runtime errors? 
- we should log an error so any unexpected subsequent processing can be diagnosed.

Reply:
> I would be keener to have an error, it is then obvious this is an out of the ordinary situation

We only do that for `CAST` because CAST explicitly defines that in the SQL standard. That's also why there's a `TRY_CAST`. We don't fail on any of the other functions (except `ELEMENT` if I saw it correctly)

Reply:
@davidradl thanks for your input

here we are talking about returning null on SQL level so I don't get what NullPointer exactly you mean, could you please elaborate on this?

In Flink there was already a functionality throwing an error (`cast`) 
Now the problem with this `cast` is that imagine you have a table with a string column however actually there are  numbers written there and we can write a query
```sql
SELECT cast(str_column to INTEGER) from MyTable;
```
now for some reason there is a string `abcd` in one of the records and `cast` is throwing an error on it. 

How are you going to even detect this row with such behavior?

with returning `null` it is much easier (`try_cast` could return `null`), we could simply write something like
```sql
SELECT * from (SELECT try_cast(str_column to INTEGER) a, str_column b from MyTable) t WHERE a is NULL;
```


also about
> runtime exceptions like errors (e.g. out of memory).

OutOfMemory is NOT a RuntimeException, it is under Errors
https://docs.oracle.com/javase/8/docs/api/java/lang/OutOfMemoryError.html
https://docs.oracle.com/javase/8/docs/api/java/lang/RuntimeException.html

Reply:
@snuyanzin @MartijnVisser thankyou very much for your feedback.

Sergey:
Sorry if I was not clear, I did not mean to imply that JAVA runtime Exceptions and errors are the same.  
you ask `How are you going to even detect this row with such behavior?` I guess this is not possible for the cast, if we had an error message that included the string that we fails to encode or decode - this could help. 

For the null pointer- I was not thinking of a java NullPointerException, just an unexpected null in subsequent processing,  maybe if this value ended up being attempted to be written to a non nullable column. 
Martjin:
I like your examples of cast and try_cast that was helpful for me. In addition to element (which does not look like a runtime exception) there are also json functions that produce errors (these also do not look like runtime errors).

It seems to me if we can produce a useful error message in the error including the string we could not encode / decode then this could be very helpful to be able to diagnose the issue. If not then returning null looks preferable. Either way I think we should log the runtime error.

Reply:
how about this:
add a WARN log before return null?